rosetta/transactions/intent.go

```diff
@@ -35,83 +36,48 @@ type Intent struct {
 	Payer    flow.Address
 	Proposer flow.Address
 
-	ReferenceBlock            flow.Identifier
-	ProposerKeySequenceNumber uint64
-	GasLimit                  uint64
+	ReferenceBlock flow.Identifier
+	SequenceNumber uint64
+	GasLimit       uint64
 }
 
-// CreateTransactionIntent creates a transaction Intent from two operations given as input.
+// DeriveIntent derives a transaction Intent from two operations given as input.
 // Specified operations should be symmetrical, a deposit and a withdrawal from two
 // different accounts. At the moment, the only fields taken into account are the
 // account IDs, amounts and type of operation.
-func (p *Parser) CreateTransactionIntent(operations []object.Operation) (*Intent, error) {
+func (p *Parser) DeriveIntent(operations []object.Operation) (*Intent, error) {
 
-	firstOpNegative := strings.HasPrefix(operations[0].Amount.Value, "-")
-	secondOpNegative := strings.HasPrefix(operations[1].Amount.Value, "-")
-
-	// Operations are invalid if both operations have negative amounts.
-	if firstOpNegative && secondOpNegative {
-
-		return nil, failure.InvalidIntent{
-			Description: failure.NewDescription("invalid operations - two deposits specified"),
-			Sender:      operations[0].AccountID.Address,
-			Receiver:    operations[1].AccountID.Address,
+	if len(operations) != 2 {
+		return nil, failure.InvalidOperations{
+			Description: failure.NewDescription("invalid number of operations"),
+			Count:       len(operations),
 		}
 	}
 
-	// Operations are also invalid if both operations have positive amounts.
-	if !firstOpNegative && !secondOpNegative {
-
-		return nil, failure.InvalidIntent{
-			Description: failure.NewDescription("invalid operations - two withdrawals specified"),
-			Sender:      operations[0].AccountID.Address,
-			Receiver:    operations[1].AccountID.Address,
-		}
-	}
+	// Sort the operations so that the send operation (negative amount) comes first.
+	sort.Slice(operations, func(i, j int) bool {
+		return operations[i].Amount.Value < operations[j].Amount.Value
+	})
 
-	// Assume the first operation is the one with the negative amount.
 	send := operations[0]
 	receive := operations[1]
 
-	// If that was not the case, switch the send and receive operations.
+	// Verify that the send operation is negative.
+	// We don't need to check the receive amount since it will be parsed as an uint64
+	// which will err on inapropriate input.
 	if !strings.HasPrefix(send.Amount.Value, "-") {
```

Comment:
This is the only place where I've left string checks. The "receive" amount will fail on parsing as uint64 if it's negative. Since we're also parsing "send" operation as an uint64, here we check if it's negative before trimming the prefix and parsing it.

Reply:
That's not the right approach, imo. The right approach is to parse it as `int64`, and then case it to `uint64`.